packages/react/src/utils/useFocusTrap.ts

```diff
@@ -0,0 +1,245 @@
+import type { ElementOrRef } from '../types';
+import { useEffect, useRef } from 'react';
+import focusable from 'focusable';
+import resolveElement from './resolveElement';
+
+type FocusTrapMetadata = {
+  targetElement: Element;
+  lastFocusedElement: HTMLElement | null;
+  suspended: boolean;
+};
+
+type FocusTrap = {
+  initialFocusElement: HTMLElement | null;
+  destroy: () => void;
+} & FocusTrapMetadata;
+
+// When multiple focus traps are created, we need to keep track of all previous traps
+// in the stack and temporarily suspend any traps created before the most recent trap
+const focusTrapStack: Array<FocusTrapMetadata> = [];
+const removeFocusTrapFromStack = (focusTrap: FocusTrapMetadata): void => {
+  const focusTrapIndex = focusTrapStack.findIndex(
+    (trap) => focusTrap.targetElement === trap.targetElement
+  );
+  focusTrapStack.splice(focusTrapIndex, 1);
+};
+
+function getActiveElement(target: Element | null): HTMLElement {
+  return (
+    ((target?.ownerDocument.activeElement as HTMLElement) ??
+      (document.activeElement as HTMLElement)) ||
+    document.body
+  );
+}
+
+function elementContains(
+  containerElement: Element,
+  targetElement: Element | null
+): boolean {
+  if (!targetElement) {
+    return false;
+  }
+
+  if (containerElement.getRootNode() === targetElement.getRootNode()) {
+    return containerElement.contains(targetElement);
+  }
+
+  let root = targetElement.getRootNode();
+  while (root && root !== containerElement.getRootNode()) {
+    if (root.nodeType === Node.DOCUMENT_FRAGMENT_NODE) {
+      // likely a shadow root, and we need to get the host
+      root = (root as ShadowRoot).host;
+    } else {
+      break;
+    }
+  }
+  return root && containerElement.contains(root);
+
+  return false;
+}
+
+function createTrapGuard(): HTMLSpanElement {
+  const trapGuard = document.createElement('span');
+  trapGuard.setAttribute('tabindex', '0');
+  trapGuard.setAttribute('aria-hidden', 'true');
+  return trapGuard;
+}
+
+function createFocusTrap(
+  targetElement: Element,
+  initialFocusElement: HTMLElement | null
+): FocusTrap {
+  const startGuard = createTrapGuard();
+  const endGuard = createTrapGuard();
```

Comment:
Part of the methodology I evaluated to switching to use trap guards included:

- Not wanting to keep track of tabbable elements (newly added or removed) within the focus trap
- Relying on native browser behavior for forward/reverse focus tab navigation
- Not wanting to account for tabIndex order when tracking elements

Given the many nuances to tracking tabbable elements within the trap, using focus guards appeared to be a much more sustainable and maintainable approach.